### ISSUE TITLE
Sending large batch of notifications cause memory overflow

### ISSUE THREAD

**Issue by sachajln** [node-apn/node-apn#578]
When trying to send large batch of notifications (for my tests I was sending a notification to 5k tokens every second, all with the same provider), node quickly consumes a lot of memory and sometimes overflows (> 1.4gb used). Is there anything I can do to fix this ? Or is it working as intended ?

For information : Ultimately I'll need to send approximately 250k push notifications every day at 9 am without risking crashing the server.

**Comment by oliversalzburg** [comment]
Do it in batches or use multiple processes

**Comment by florianreinhart** [comment]
Yes, I would use multiple processes.

You might also consider using local notifications on iOS if the content of your notifications is not dynamically generated in the backend.

**Comment by deepakc** [comment]
+1 
Seeing this memory issue only after the http2 implementation. The older version using legacy apns is not having this memory issue.

**Comment by jpike88** [comment]
This issue should be closed or merged into the http2 memory leak issue... they're clearly linked.

Also @sachajln this testing code, can you share it?

**Comment by dpmooney** [comment]
I'm not @sachajln, but having the same issues w/ 10K notifications.  I made a simple wrapper for node-apn that chunks the devices into 150 notification blocks and sends them out.  Three issues appear: 

1. Memory grows quickly
2. I receive errors for maxlisteners: `(node:54626) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 wakeup listeners added. Use emitter.setMaxListeners() to increase limit`
3. After notifications have sent and I've received all send-statuses, it takes about a minute for node to drop back to console.

Here's a snippet of my code:

```javascript
    var device_array = devices.chunk(chunkSize) // chunksize being 150 currently
    device_array.forEach(function(devices){
      promises.push(tnp.provider.send(notification.body, devices)
        .then(function(result){
          return result
        })
      )
    })

    return Promise.all(promises).then(function(res){
      return res
    })
```

**Comment by andreialecu** [comment]
Related, and possible solution: https://github.com/node-apn/node-apn/issues/557#issuecomment-340812136
